### ISSUE TITLE
Kotlin SAM overload resolution ambiguity between Consumer and ThrowingConsumer

### ISSUE THREAD

**Issue by ansgarkonermann** [assertj/assertj-core#2357]
#### Summary
When using AssertJ Core 3.21.0 from Kotlin, an overload resolution ambiguity occurrs for methods having overloads which accept both Consumer and ThrowingConsumer.

Methods affected (list noncomprehensive)
```
org.assertj.core.api.AbstractAssert#satisfies(java.util.function.Consumer<? super ACTUAL>)
org.assertj.core.api.ObjectEnumerableAssert#allSatisfy(org.assertj.core.api.ThrowingConsumer<? super ELEMENT>)
``` 
#### Example

Worked before (typical syntax for Kotlin calling a method which takes a functional interface AKA "SAM")

```kotlin
assertThat(bounds).satisfies { // worked pre-3.21.0
    assertThat(it.startInclusive) // ... any detail assert statement
}
```

Wordy workaround:
```kotlin
assertThat(bounds).satisfies(Consumer {
    assertThat(it.startInclusive) // ... any detail assert statement
})
```
This overload ambiguity is (also) an issue of kotlin language design, cf. https://youtrack.jetbrains.com/issue/KT-17765

**Comment by joel-costigliola** [comment]
@ansgarkonermann thanks for the investigation, we'll have to wait for the Kotlin fix for this one since there is a workaround.

**Comment by piefel** [comment]
I do not think that this _can_ be fixed in Kotlin. It is not a bug in the language. To identify SAM interfaces, we can take arguments to the called function and the result type into consideration. If after that, more than one interface matches, we are out of luck. Now _exceptions_ are simply not taken into consideration here – and this is by design.

I somehow seem to remember a similar issue with `.extracting()` – was there someting like a `ThrowingFunction` once upon a time?

BTW, the work-around is super-ugly with our standard formatter, which will put the `Consumer {` on a line by itself and requires another level of indentation. I wrote a `satifisfiesSoftly()` extension function just to avoid this (and add softness on top). I would very much wish for AssertJ to stay compatible with Kotlin. I know there are other assertion libraries, but they are just not up to par.

**Comment by joel-costigliola** [comment]
@elizarov is there a plan to address https://youtrack.jetbrains.com/issue/KT-17765?

**Comment by alex-arana** [comment]
I've created the following workaround in my tests. For the record, I still believe a more sensible approach would have been to name the new API `satisfiesThrowable` or similar.

```kotlin
/**
 * Convenience _AssertJ_ method used to verify that the receiver's actual value satisfies the specified test requirements
 * as a lambda [block].
 *
 * *NOTE*: As of version `3.21.0` [AssertJ](https://assertj.github.io/doc/) [satisfies](https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/AbstractAssert.html#satisfies(java.util.function.Consumer))
 *   is irretrievably broken in Kotlin due to the rather unfortunate introduction of a method override using a [ThrowingConsumer](https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/ThrowingConsumer.html)
 *   parameter which makes both methods indistinguishable while using single-abstraction-method (SAM) semantics.
 *
 * @param block Lambda which accepts a non-nullable instance of the _actual_ asserted object as an input parameter
 */
fun <T : Any> Assert<*, T>.satisfiesRequirements(block: (T) -> Unit): Assert<*, *> {
    return this.satisfies(object : Condition<T>() {
        override fun matches(value: T): Boolean {
            block(value)
            return true
        }
    })
}
```

**Comment by joel-costigliola** [comment]
https://youtrack.jetbrains.com/issue/KT-17765 has been assigned so maybe this will be fixed.

**Comment by joel-costigliola** [comment]
https://youtrack.jetbrains.com/issue/KT-17765 is now fixed and is targeted for 1.6.20 🎉

**Comment by joel-costigliola** [comment]
Closing this ticket since it's addressed in Kotlin.

**Comment by 3flex** [comment]
Though KT-17765 seems fixed in Kotlin 1.6.20-M1 it requires using experimental language version 1.7, e.g.:

```kotlin
tasks.compileTestKotlin {
    kotlinOptions.languageVersion = "1.7"
}
```

**Comment by bananer** [comment]
I could not get this to compile even with Kotlin 1.6.20 and setting languageVersion to 1.7

Here's my workaround, a bit shorter than the version by @alex-arana :
```kotlin
import org.assertj.core.api.AbstractIterableAssert
import java.util.function.Consumer

fun <E : Any?, I : Iterable<E>> AbstractIterableAssert<*, I, E, *>.allSatisfyKt(requirements: Consumer<E>): AbstractIterableAssert<*, *, *, *> {
    return this.allSatisfy(requirements)
}
```

**Comment by alex-arana** [comment]
> Though KT-17765 seems fixed in Kotlin 1.6.20-M1 it requires using experimental language version 1.7, e.g.:
> 
> ```kotlin
> tasks.compileTestKotlin {
>     kotlinOptions.languageVersion = "1.7"
> }
> ```

I've tested using experimental language version 1.7 as suggested and it works.. kinda
![idea](https://user-images.githubusercontent.com/2107295/162099589-fa99ba1e-48a3-4016-b4c1-ca7f507e78f4.jpg)

As can be seen in the attached image, braces around the `Consumer` lambda are still required..

**Comment by alex-arana** [comment]
> > As can be seen in the attached image, braces around the Consumer lambda are still required..
> 
> Is that just in the IDE or do you also get compilation errors?
> 
> If your IDE Kotlin plugin version is out of date that might result in what you're seeing, I assume you would need IDE plugin version 1.6.20 or greater.

Compilation error occurs during Gradle build as well.. I am using Kotlin Plugin `213-1.6.20-release-275-IJ6777.52`

![Screen Shot 2022-04-11 at 11 38 53 am](https://user-images.githubusercontent.com/2107295/162651449-ade228f1-3c36-4960-bec6-826278dc217b.png)

**Comment by oucem** [comment]
Same case as @alex-arana with maven plugin

**Comment by johannesknauft** [comment]
I had the same problem after I updated spring boot. For me, the problem was solved by downgrade assertj-core to version `3.20.2` (`testImplementation("org.assertj:assertj-core:3.20.2")`). I guess that's not a long-term solution, but a workaround for now.

**Comment by cy6erGn0m** [comment]
Note that having experimental options is not the way for many enterprise projects. So having the fix will take years do get actually delivered, so what we supposed to do before?

**Comment by joel-costigliola** [comment]
Can you guys raise another Kotlin issue for this one to be properly addressed?

On our side, I would be happy to make some changes if they are backward compatible but I doubt this is possible so I think the right solution is to ask the Kotlin project to be more java compatible.

**Comment by piefel** [comment]
Well, I would really wish AssertJ was more Kotlin aware. It is not the first time things broke. Java interop in Kotlin is great, but can only ever go so far without sacrificing the language. 

⁣Gesendet mit TypeApp ​

**Comment by joel-costigliola** [comment]
The AssertJ team doesn't have the skills or the capacity to ensure AssertJ is Kotlin compatible at the moment, we can make some tweaks but it's java first.

**Comment by pkubowicz** [comment]
Kotlin 1.7 works with AssertJ 3.21.0. It won't however work with AssertJ 3.22.0, as a new problem appears:

```plain
None of the following functions can be called with the arguments supplied: 
public final fun satisfies(vararg p0: Consumer<in String!>!): ObjectAssert<String!>! defined in org.assertj.core.api.ObjectAssert
public final fun satisfies(vararg p0: ThrowingConsumer<in String!>!): ObjectAssert<String!>! defined in org.assertj.core.api.ObjectAssert
public open fun satisfies(p0: Condition<in String!>!): ObjectAssert<String!>! defined in org.assertj.core.api.ObjectAssert
```

Note this is the same `satisfies` method but a different error message than the one originally reported here. Also the reason is different: Kotlin compiler is unable to properly handle vararg use of Consumer and ThrowingConsumer. Reported as a separate issue: https://youtrack.jetbrains.com/issue/KT-52942. If you found my comment because you experience this new problem, upvote the issue in Jetbrains' issue tracker.

**Comment by spyro2000** [comment]
See also https://youtrack.jetbrains.com/issue/KT-53113/Kotlin-AssertJ-tests-compile-in-1621-but-not-in-17

 It's an disturbing blame game between AssertJ and Kotlin with fingerpointing on both sides but still no solution.
AssertJ (as Bundled with Spring Boot 2.6+) does simply not work with Kotlin currently.

**Comment by scordio** [comment]
> See also https://youtrack.jetbrains.com/issue/KT-53113/Kotlin-AssertJ-tests-compile-in-1621-but-not-in-17

[KT-53113](https://youtrack.jetbrains.com/issue/KT-53113) is actually related to #2439.

**Comment by scordio** [comment]
Alright, let's reopen this one.

I think it's too late to introduce further breaking changes in version 3 for this topic, but version 4 may be the right time to rethink how to better support Kotlin.

**Comment by scordio** [comment]
As also mentioned in [KT-52942](https://youtrack.jetbrains.com/issue/KT-52942/):

> The point is that AssertJ makes extensive use of Java Generic Types and Kotlin does not seem to support all use cases for generic types allowed by Java.

**Comment by scordio** [comment]
[KT-17765](https://youtrack.jetbrains.com/issue/KT-17765) has been closed as a duplicate of [KT-7052](https://youtrack.jetbrains.com/issue/KT-7052), the latter mentioning:

> Unfortunately, the issue is much more global than it seems, so it's unlikely it will be fixed in the ongoing release.
